File: src/lib/prng.ts
// Seeded random numbers so the generated dataset is identical on every machine.

export function mulberry32(seed: number) {
  let a = seed >>> 0;
  return function next() {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export type Rng = ReturnType<typeof mulberry32>;

export function randInt(rng: Rng, lo: number, hi: number) {
  return lo + Math.floor(rng() * (hi - lo + 1));
}

export function pick<T>(rng: Rng, arr: readonly T[]): T {
  return arr[Math.floor(rng() * arr.length)];
}

export function normal(rng: Rng, mean = 0, sd = 1) {
  const u = Math.max(rng(), 1e-12);
  const v = rng();
  return mean + sd * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

export function poisson(rng: Rng, lambda: number) {
  if (lambda <= 0) return 0;
  if (lambda > 30) return Math.max(0, Math.round(normal(rng, lambda, Math.sqrt(lambda))));
  const L = Math.exp(-lambda);
  let k = 0;
  let p = 1;
  do {
    k++;
    p *= rng();
  } while (p > L);
  return k - 1;
}

export function binomial(rng: Rng, n: number, p: number) {
  if (n <= 0 || p <= 0) return 0;
  if (n > 60) return Math.min(n, Math.max(0, Math.round(normal(rng, n * p, Math.sqrt(n * p * (1 - p))))));
  let k = 0;
  for (let i = 0; i < n; i++) if (rng() < p) k++;
  return k;
}

File: backend/supply/prng.py
"""Seeded random numbers, so the generated dataset is the same on every machine.

mulberry32 in 32-bit integer arithmetic. It gives the same sequence the original TypeScript
generator gave, so a planning date produces the same dataset as before the port.
"""

import math
from collections.abc import Sequence

M32 = 0xFFFFFFFF


def _imul(a: int, b: int) -> int:
    return (a * b) & M32


class Rng:
    def __init__(self, seed: int) -> None:
        self.a = seed & M32

    def __call__(self) -> float:
        self.a = (self.a + 0x6D2B79F5) & M32
        t = self.a
        t = _imul(t ^ (t >> 15), t | 1)
        t ^= (t + _imul(t ^ (t >> 7), t | 61)) & M32
        return ((t ^ (t >> 14)) & M32) / 4294967296


def js_round(x: float) -> int:
    """Round half up, as JavaScript's Math.round does. Python's round() rounds half to even."""
    return math.floor(x + 0.5)


def rand_int(rng: Rng, lo: int, hi: int) -> int:
    return lo + math.floor(rng() * (hi - lo + 1))


def pick[T](rng: Rng, arr: Sequence[T]) -> T:
    return arr[math.floor(rng() * len(arr))]


def normal(rng: Rng, mean: float = 0.0, sd: float = 1.0) -> float:
    u = max(rng(), 1e-12)
    v = rng()
    return mean + sd * math.sqrt(-2 * math.log(u)) * math.cos(2 * math.pi * v)


def poisson(rng: Rng, lam: float) -> int:
    if lam <= 0:
        return 0
    if lam > 30:
        return max(0, js_round(normal(rng, lam, math.sqrt(lam))))
    limit = math.exp(-lam)
    k = 0
    p = 1.0
    while True:
        k += 1
        p *= rng()
        if p <= limit:
            break
    return k - 1


def binomial(rng: Rng, n: int, p: float) -> int:
    if n <= 0 or p <= 0:
        return 0
    if n > 60:
        return min(n, max(0, js_round(normal(rng, n * p, math.sqrt(n * p * (1 - p))))))
    return sum(1 for _ in range(n) if rng() < p)
